Implement trusted types integration with DOM node conversion APIs

lukewarlow opened the pull request:
#### bb9ca28dcac916cb783096b485f9c589b147f21d
<pre>
Implement trusted types integration with DOM node conversion APIs
<a href="https://bugs.webkit.org/show_bug.cgi?id=270435">https://bugs.webkit.org/show_bug.cgi?id=270435</a>

Reviewed by Ryosuke Niwa.

Update the &quot;converting nodes into a node&quot; algorithm to integrate Trusted Types protection.

* LayoutTests/imported/w3c/web-platform-tests/trusted-types/Node-multiple-arguments-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-Node-multiple-arguments-expected.txt:
* Source/WebCore/dom/ChildNode.idl:
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::append):
(WebCore::ContainerNode::prepend):
(WebCore::ContainerNode::replaceChildren):
* Source/WebCore/dom/ContainerNode.h:
* Source/WebCore/dom/Node.cpp:
(WebCore::nodeSetPreTransformedFromNodeOrStringOrTrustedScriptVector):
(WebCore::Node::convertNodesOrStringsOrTrustedScriptsIntoNode):
(WebCore::Node::convertNodesOrStringsOrTrustedScriptsIntoNodeVector):
(WebCore::Node::before):
(WebCore::Node::after):
(WebCore::Node::replaceWith):
(WebCore::nodeSetPreTransformedFromNodeOrStringVector): Deleted.
(WebCore::Node::convertNodesOrStringsIntoNode): Deleted.
(WebCore::Node::convertNodesOrStringsIntoNodeVector): Deleted.
* Source/WebCore/dom/Node.h:
* Source/WebCore/dom/ParentNode.idl:
* Source/WebCore/dom/TrustedType.cpp:
(WebCore::processNodeOrStringAsTrustedType):
* Source/WebCore/dom/TrustedType.h:

Canonical link: <a href="https://commits.webkit.org/276942@main">https://commits.webkit.org/276942@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e3dee18c95721eb4fb02e1e897ed96009c6bd73a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46210 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25341 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48795 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48882 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42251 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48517 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29687 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22796 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37762 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46788 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22422 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39852 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18973 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19811 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40950 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4254 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/39443 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42519 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41300 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50690 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/45684 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21206 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17703 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44952 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22506 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/40009 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43867 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10232 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22865 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/52835 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22200 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10816 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

lukewarlow commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -595,18 +598,42 @@ static RefPtr<Node> firstFollowingSiblingNotInNodeSet(Node& context, const HashS
+    NodeVector nodes;
+    for (auto& variant : vector) {
+        RefPtr<Text> text;
+        if (std::holds_alternative<String>(variant))
```
Feels like there's probably a cleaner way to write this block

rwlbuis replied:
Is it possible to keep the WTF::switchOn structure and put the added functionality for TT in some helper function/method?

lukewarlow replied:
I don't know that I can keep the WTF::switchOn structure, but I can definitely abstract some of this to a helper function in TrustedType.cpp. Will push up and see what you think.

lukewarlow commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -621,17 +648,36 @@ ExceptionOr<RefPtr<Node>> Node::convertNodesOrStringsIntoNode(FixedVector<NodeOr
-        if (std::holds_alternative<String>(variant)) {
-            nodeVector.append(Text::create(document, WTFMove(std::get<String>(variant))));
+    nodeVector.reserveInitialCapacity(vector.size());
+    for (auto& variant : vector) {
```
ditto

rwlbuis replied:
See above.

rwlbuis commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -621,17 +635,23 @@ ExceptionOr<RefPtr<Node>> Node::convertNodesOrStringsIntoNode(FixedVector<NodeOr
-    for (auto& variant : nodeOrStringVector) {
-        if (std::holds_alternative<String>(variant)) {
-            nodeVector.append(Text::create(document, WTFMove(std::get<String>(variant))));
+    nodeVector.reserveInitialCapacity(vector.size());
```
I think this is not properly doing step 2 before doing the rest of the algorithm:
https://whatpr.org/dom/1258.html#converting-nodes-into-a-node

Also should there not be a feature flag for the case TT is not enabled somewhere here?

lukewarlow replied:
The feature flags are handled in the TT helper functions.

lukewarlow replied:
Have left a comment in the helper function explaining how step 2 is covered.

rwlbuis replied:
Where do the feature flags come into play? It seems with this change String is doing an extra check for parent being script, and Node is doing an extra check if it is a TrustedScript, despite the flag being turned off?

lukewarlow replied:
Okay yeah extra code is ran in the situation where the flag is off. But there's no behaviour change, the flag is in trustedTypeCompliantString so that function call will just return the string unchanged.

I guess I should add a feature flag check in higher up to skip those checks?

rwlbuis replied:
Right, I would aim to avoid any speed regressions when TT is tuned off.

lukewarlow replied:
Added flag checks higher up the call stack

lukewarlow commented on Source/WebCore/dom/TrustedType.cpp:
```diff
@@ -207,4 +210,28 @@ ExceptionOr<String> requireTrustedTypesForPreNavigationCheckPasses(ScriptExecuti
+{
+    RefPtr<Text> text;
+    if (std::holds_alternative<String>(variant))
+        text = Text::create(document, WTFMove(std::get<String>(variant)));
```
@rwlbuis it's not quite as the spec is written (because WebKit isn't quite how the spec is written already) but this line should be effectively doing step 2.

And then the text node is processed below. Which handles Original Strings and Original Text Nodes in one go.

rwlbuis commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -621,17 +635,23 @@ ExceptionOr<RefPtr<Node>> Node::convertNodesOrStringsIntoNode(FixedVector<NodeOr
+
+        RefPtr<Text> textNode = textNodeHolder.releaseReturnValue();
+
+        if (textNode) {
```
I think it nicer to combine these:
if ( RefPtr<Text> textNode = ... )

rniwa commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -595,18 +598,33 @@ static RefPtr<Node> firstFollowingSiblingNotInNodeSet(Node& context, const HashS
-    });
+    NodeVector nodes;
+    for (auto& variant : vector) {
+        if (document->scriptExecutionContext()->settingsValues().trustedTypesEnabled) {
```
We should read this setting value once, not in every iteration.

lukewarlow replied:
Ah good catch yes

rniwa commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -621,19 +639,29 @@ ExceptionOr<RefPtr<Node>> Node::convertNodesOrStringsIntoNode(FixedVector<NodeOr
-        if (std::holds_alternative<String>(variant)) {
+    nodeVector.reserveInitialCapacity(vector.size());
+    for (auto& variant : vector) {
+        if (document->scriptExecutionContext()->settingsValues().trustedTypesEnabled) {
```
Ditto.

rniwa commented on Source/WebCore/dom/TrustedType.cpp:
```diff
@@ -209,4 +212,28 @@ ExceptionOr<String> requireTrustedTypesForPreNavigationCheckPasses(ScriptExecuti
+    }
+
+    if (text) {
+        if (is<HTMLScriptElement>(parent)) {
```
Can we add UNLIKELY here?

lukewarlow committed:
Implement trusted types integration with DOM node conversion APIs
https://bugs.webkit.org/show_bug.cgi?id=270435

Reviewed by Ryosuke Niwa.

Update the "converting nodes into a node" algorithm to integrate Trusted Types protection.

* LayoutTests/imported/w3c/web-platform-tests/trusted-types/Node-multiple-arguments-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-Node-multiple-arguments-expected.txt:
* Source/WebCore/dom/ChildNode.idl:
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::append):
(WebCore::ContainerNode::prepend):
(WebCore::ContainerNode::replaceChildren):
* Source/WebCore/dom/ContainerNode.h:
* Source/WebCore/dom/Node.cpp:
(WebCore::nodeSetPreTransformedFromNodeOrStringOrTrustedScriptVector):
(WebCore::Node::convertNodesOrStringsOrTrustedScriptsIntoNode):
(WebCore::Node::convertNodesOrStringsOrTrustedScriptsIntoNodeVector):
(WebCore::Node::before):
(WebCore::Node::after):
(WebCore::Node::replaceWith):
(WebCore::nodeSetPreTransformedFromNodeOrStringVector): Deleted.
(WebCore::Node::convertNodesOrStringsIntoNode): Deleted.
(WebCore::Node::convertNodesOrStringsIntoNodeVector): Deleted.
* Source/WebCore/dom/Node.h:
* Source/WebCore/dom/ParentNode.idl:
* Source/WebCore/dom/TrustedType.cpp:
(WebCore::processNodeOrStringAsTrustedType):
* Source/WebCore/dom/TrustedType.h:

Canonical link: https://commits.webkit.org/276942@main
